source/logic.rst

```diff
@@ -0,0 +1,878 @@
+        fn_ub (λ x, f x + g x) (a + b) :=
+    λ x, add_le_add (hfa x) (hgb x)
+
+    -- BEGIN
+    example (ubf : fn_has_ub f) (ubg : fn_has_ub g) :
+      fn_has_ub (λ x, f x + g x) :=
+    begin
+      rcases ubf with ⟨a, ubfa⟩,
+      rcases ubg with ⟨b, ubfb⟩,
+      exact ⟨a + b, fn_ub_add ubfa ubfb⟩
+    end
+
+    example : fn_has_ub f → fn_has_ub g →
+      fn_has_ub (λ x, f x + g x) :=
+    begin
+      rintros ⟨a, ubfa⟩ ⟨b, ubfb⟩,
+      exact ⟨a + b, fn_ub_add ubfa ubfb⟩
+    end
+    -- END
+
+In fact, Lean also supports a pattern-matching lambda
+in expressions and proof terms:
+
+.. code-block:: lean
+
+    import data.real.basic
+
+    def fn_ub (f : ℝ → ℝ) (a : ℝ) : Prop := ∀ x, f x ≤ a
+    def fn_lb (f : ℝ → ℝ) (a : ℝ) : Prop := ∀ x, a ≤ f x
+
+    def fn_has_ub (f : ℝ → ℝ) := ∃ a, fn_ub f a
+    def fn_has_lb (f : ℝ → ℝ) := ∃ a, fn_lb f a
+
+    variables {f g : ℝ → ℝ}
+
+    theorem fn_ub_add {f g : ℝ → ℝ} {a b : ℝ}
+        (hfa : fn_ub f a) (hgb : fn_ub g b) :
+        fn_ub (λ x, f x + g x) (a + b) :=
+    λ x, add_le_add (hfa x) (hgb x)
+
+    -- BEGIN
+    example : fn_has_ub f → fn_has_ub g →
+      fn_has_ub (λ x, f x + g x) :=
+    λ ⟨a, ubfa⟩ ⟨b, ubfb⟩, ⟨a + b, fn_ub_add ubfa ubfb⟩
+    -- END
+
+These are power-user moves, and there is no harm
+in favoring the use of ``cases`` until you are more comfortable
+with the existential quantifier.
+But we will come to learn that all of these tools,
+including ``cases``, use, and the anonymous constructors,
+are like Swiss army knives when it comes to theorem proving.
+They can be used for a wide range of purposes,
+not just for unpacking exists statements.
+
+To illustrate one way that ``rcases`` can be used,
+we prove an old mathematical chestnut:
+if two integers ``x`` and ``y`` can each be written as
+a sum of two squares,
+then so can their product, ``x * y``.
+In fact, the statement is true for any commutative
+ring, not just the integers.
+In the next example, ``rcases`` unpacks two existential
+quantifiers at once.
+We then provide the magic values needed to express ``x * y``
+as a sum of squares as a list to the ``use`` statement,
+and we use ``ring`` to verify that they work.
+
+.. code-block:: lean
+
+    import tactic
+
+    variables {α : Type*} [comm_ring α]
+
+    def sos (x : α) := ∃ a b, x = a^2 + b^2
+
+    theorem sos_mul {x y : α} (sosx : sos x) (sosy : sos y) : sos (x * y) :=
+    begin
+      rcases sosx with ⟨a, b, xeq⟩,
+      rcases sosy with ⟨c, d, yeq⟩,
+      use [a*c - b*d, a*d + b*c],
+      rw [xeq, yeq], ring
+    end
+
+As with the universal quantifier,
+you can find existential quantifiers hidden all over
+if you know how to spot them.
+For example, divisibility is implicitly an "exists" statement.
+
+.. code-block:: lean
+
+    import tactic
+
+    variables {a b c : ℕ}
+
+    -- BEGIN
+    example (divab : a ∣ b) (divbc : b ∣ c) : a ∣ c :=
+    begin
+      cases divab with d beq,
+      cases divbc with e ceq,
+      rw [ceq, beq],
```

Comment:
Again, after seeing this pattern `cases ..., rw` a couple of times, our readers may be ready to learn the power-user way. This is not a computer scientist trick, it really matches how we think about it: "a ∣ b hence we can replace b by a*d for some d" seen as one reasoning step.

Reply:
Good idea -- I'll introduce it. (I like all your other suggestions as well.)